File: sdk/typescript/src/retry-policy.ts
/**
 * Per-error-type retry configuration.
 *
 * @module
 */

export interface RetryPolicyOptions {
  rateLimitRetries?: number;
  timeoutRetries?: number;
  connectionErrorRetries?: number;
  internalServerErrorRetries?: number;
  badGatewayRetries?: number;
  serviceUnavailableRetries?: number;
  gatewayTimeoutRetries?: number;
}

const STATUS_FIELDS: Record<number, keyof RetryPolicyOptions> = {
  429: "rateLimitRetries",
  500: "internalServerErrorRetries",
  502: "badGatewayRetries",
  503: "serviceUnavailableRetries",
  504: "gatewayTimeoutRetries",
};

const DEFAULTS: Required<RetryPolicyOptions> = {
  rateLimitRetries: 3,
  timeoutRetries: 2,
  connectionErrorRetries: 2,
  internalServerErrorRetries: 2,
  badGatewayRetries: 1,
  serviceUnavailableRetries: 1,
  gatewayTimeoutRetries: 1,
};

export class RetryPolicy {
  private _opts: Required<RetryPolicyOptions>;

  constructor(opts: RetryPolicyOptions = {}) {
    this._opts = { ...DEFAULTS, ...opts };
  }

  getRetriesForStatus(statusCode: number): number {
    const field = STATUS_FIELDS[statusCode];
    if (field) return this._opts[field] as number;
    return 0;
  }

  getRetriesForConnectionError(): number {
    return this._opts.connectionErrorRetries;
  }

  getRetriesForTimeout(): number {
    return this._opts.timeoutRetries;
  }
}

File: sdk/python/src/agentcc/_retry_policy.py
"""Per-exception-type retry policy configuration."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class RetryPolicy:
    """Per-exception-type retry configuration.

    Instead of a flat ``max_retries`` for all error types, this allows
    fine-grained control over how many retries each error category gets.

    Usage::

        from agentcc import RetryPolicy

        policy = RetryPolicy(RateLimitErrorRetries=5, TimeoutRetries=3)
        client = AgentCC(retry_policy=policy, api_key="sk-...", base_url="...")
    """

    RateLimitErrorRetries: int = 2
    TimeoutRetries: int = 2
    ConnectionErrorRetries: int = 2
    InternalServerErrorRetries: int = 1
    BadGatewayRetries: int = 1
    ServiceUnavailableRetries: int = 1
    GatewayTimeoutRetries: int = 1

    def get_retries_for_status(self, status_code: int) -> int:
        """Return the max retries for a given HTTP status code."""
        mapping = {
            429: self.RateLimitErrorRetries,
            500: self.InternalServerErrorRetries,
            502: self.BadGatewayRetries,
            503: self.ServiceUnavailableRetries,
            504: self.GatewayTimeoutRetries,
        }
        return mapping.get(status_code, 0)

    def get_retries_for_connection_error(self) -> int:
        """Return the max retries for connection errors."""
        return self.ConnectionErrorRetries

    def get_retries_for_timeout(self) -> int:
        """Return the max retries for timeout errors."""
        return self.TimeoutRetries
